crypto/ml_kem/ml_kem.c

```diff
@@ -0,0 +1,1567 @@
+ * (DEGREE) is divisible by 8, the individual vector entries will always fill a
+ * whole number of bytes, so we do not need to worry about bit packing here.
+ */
+static void vector_encode(uint8_t *out, const scalar *a, int bits, int rank)
+{
+    int stride = bits * DEGREE / 8;
+
+    for (; rank-- > 0; out += stride)
+        scalar_encode(out, a++, bits);
+}
+
+/*
+ * Decodes 32*|rank|*|bits| bytes from |in| into |out|. It returns one on
+ * success or zero if any parsed value is >= |ML_KEM_PRIME|.
+ */
+static int vector_decode(scalar *out, const uint8_t *in, int bits, int rank)
+{
+    int stride = bits * DEGREE / 8;
+
+    for (; rank-- > 0; in += stride)
+        if (!scalar_decode(out++, in, bits))
+            return 0;
+    return 1;
+}
+
+/* vector_encode, specialised to bits == 12. */
+static void vector_encode_12(uint8_t *out, const scalar *a, int rank)
+{
+    int stride = 3 * DEGREE / 2;
+
+    for (; rank-- > 0; out += stride)
+        scalar_encode_12(out, a++);
+}
+
+/* vector_decode, specialised to bits == 12. */
+static __owur
+int vector_decode_12(scalar *out, const uint8_t *in, int rank)
+{
+    int stride = 3 * DEGREE / 2;
+
+    for (; rank-- > 0; in += stride)
+        if (!scalar_decode_12(out++, in))
+            return 0;
+    return 1;
+}
+
+static void vector_compress(scalar *a, int bits, int rank)
+{
+    while (rank-- > 0)
+        scalar_compress(a++, bits);
+}
+
+static void vector_decompress(scalar *a, int bits, int rank)
+{
+    while (rank-- > 0)
+        scalar_decompress(a++, bits);
+}
+
+static void inner_product(scalar *out, const scalar *lhs, const scalar *rhs,
+                          int rank)
+{
+    scalar_mult(out, lhs, rhs);
+    while (--rank > 0)
+        scalar_mult_add(out, ++lhs, ++rhs);
+}
+
+/* In-place NTT transform of a vector */
+static void vector_ntt(scalar *a, int rank)
+{
+    while (rank-- > 0)
+        scalar_ntt(a++);
+}
+
+/* In-place inverse NTT transform of a vector */
+static void vector_inverse_ntt(scalar *a, int rank)
+{
+    while (rank-- > 0)
+        scalar_inverse_ntt(a++);
+}
+
+static void
+matrix_mult(scalar *out, const scalar *m, const scalar *a, int rank)
+{
+    const scalar *ar;
+    int i, j;
+
+    for (i = rank; i-- > 0; ++out) {
+        scalar_mult(out, m++, ar = a);
+        for (j = rank - 1; j > 0; --j)
+            scalar_mult_add(out, m++, ++ar);
+    }
+}
+
+static void
+matrix_mult_transpose(scalar *out, const scalar *m, const scalar *a, int rank)
+{
+    const scalar *mc = m, *mr, *ar;
+    int i, j;
+
+    for (i = rank; i-- > 0; ++out)  {
+        scalar_mult(out, mr = mc++, ar = a);
+        for (j = rank; --j > 0; )
+            scalar_mult_add(out, (mr += rank), ++ar);
+    }
+}
+
+/*
+ * Generates a secret vector by using |cbd| with the given seed to generate
+ * scalar elements and incrementing |counter| for each slot of the vector.
+ */
+static __owur
+int gencbd_vector(scalar *out, cbd_t cbd, uint8_t *counter,
```

Comment:
Its not obvious that cbd == function
cbd_t should be uppercase.. Also change name to be more verbose..
existing example is OSSL_CALBACK *param_cb